crypto/rsa/rsa_ossl.c

```diff
@@ -747,6 +818,67 @@ static int rsa_ossl_mod_exp(BIGNUM *r0, const BIGNUM *I, RSA *rsa, BN_CTX *ctx)
     if (!BN_add(r0, r1, m1))
         goto err;
 
+    /* add m_i to m in multi-prime case */
+    if (primes > 0) {
+        BIGNUM *pr2 = BN_new();
+
+        if (pr2 == NULL)
+            goto err;
+
+        for (i = 0; i < primes; i++) {
+            pinfo = sk_RSA_PRIME_INFO_value(rsa->prime_infos, i);
+            if (!BN_sub(r1, m[i], r0)) {
+                BN_free(pr2);
+                goto err;
+            }
+
+            /*
+             * This will help stop the size of r0 increasing, which does affect
+             * the multiply if it optimised for a power of 2 size
+             */
+            if (BN_is_negative(r1)) {
```

Comment:
I don't follow above comment. For starters note that r1 is always negative, because r0 > m[i], r0 is wider than m[i] *and* pinfo->r. In fact I reckon that following addition can be omitted...

Reply:
Yes, the comment is misleading because it's copied from the 2-prime case and should be fixed. But despite the comment, you can't guarantee m[i] must be smaller than r0. In fact, m[i] is computed as:
```
m_i = c^(d_i) mod r_i (i >= 3)
```
so m[i] is smaller than r_i (the pinfo->r), and r0 is computed as (let's take i = 3):
```
r0 = c ^ dQ mod q + q * ((c ^ dP mod p - c ^ dQ mod q) * qInv mod q)
r0 = x + yq ( x < q and y < q)
```
So, if I understand correctly, m[i] is less than r_i (pinfo->r) and r0 is some multiples of q. So (m[i] - r0) will be positive if r_i is very large. For other primes (r4, r5, r6...), it's the same logic here. This could not happen for an OpenSSL generated key because the prime lengths are divided equally, but for other non-OpenSSL generated key, it's not guaranteed.

Reply:
Note r0 is growing in size with each iteration. At first loop iteration r0 is approximately as wide as p*q, so that for m[i] - r0 to be positive, 1st excess prime factor has to be wider/larger than p*q. And going further nth excess factor has to be wider/larger than multiple of all previous factors. But it has no effect on argument. Once again, BN_add can be omitted (and so is check for negativity). Let's say somebody generated odd-looking key with ever increasing prime factors (\*), so that result m[i] - r0 is always positive. So BN_add is not executed. But assertion is that it's redundant when result is negative or in other words when key is "normal"...

(\*) BTW, how secure that it be? Does one want to use it? Maybe it should be even rejected?

Reply:
I don't quite get it. How about positive first and the negative later, for instance, a very large 3rd prime and a very small 4th prime...

Reply:
More important: why do you think the BN_add should be omitted?

Reply:
Ahh, I get your point...

Reply:
So far I've been saying that it *could* be omitted. Only when we figure out that it's in fact true can we discuss if it *should* be omitted. So we start with why it *could* be omitted. What's the purpose of that addition in two-prime case? To actually reduce intermediate result of subtraction modulo p. As optimization! But does it actually matter if result is reduced? No, because it's explicitly reduced later on with BN_mod and conditional addition. With this alone it can be omitted as redundant, because said optimization doesn't apply when you get to excess primes, not in "normal" case when all prime factors are of approximately same width. Because in *normal* case inputs in the said loop are grossly uneven with r0 being *times* wider that m[i]. So that implied part of initial question also was if result is actually reducible by addition. In "normal" case result after addition will still remain so to say as unreduced as before addition, and it would be up to later BN_mod [and conditional addition] to tie things up.

Reply:
Yes, I agree with this. So now the only thing needs to be considered is if this piece of code should be there to cover some very rare *unnormal* keys?

Reply:
There is no point to optimize for abnormal cases. It's sufficient to say that later BN_mod [and conditional addition] do the right thing in either case, while in *normal* case the operation is actually redundant.

Reply:
Hmmm, this sounds reasonable